pennylane/templates/subroutines/all_singles_doubles.py

```diff
@@ -0,0 +1,181 @@
+        singles (Sequence[Sequence]): sequence of lists with the indices of the two qubits
+            the :class:`~.pennylane.SingleExcitation` operations act on
+        doubles (Sequence[Sequence]): sequence of lists with the indices of the four qubits
+            the :class:`~.pennylane.DoubleExcitation` operations act on
+
+    .. UsageDetails::
+
+        Notice that:
+
+        #. The number of wires has to be equal to the number of spin orbitals included in
+           the active space.
+
+        #. The single and double excitations can be generated with the function
+           :func:`~.excitations`. See example below.
+
+        An example of how to use this template is shown below:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            import numpy as np
+
+            electrons = 2
+            qubits = 4
+
+            # Define the HF state
+            hf_state = qml.qchem.hf_state(electrons, qubits)
+
+            # Generate all single and double excitations
+            singles, doubles = qml.qchem.excitations(electrons, qubits)
+
+            # Define the device
+            dev = qml.device('default.qubit', wires=qubits)
+
+            wires = range(qubits)
+
+            @qml.qnode(dev)
+            def circuit(weights, hf_state, singles, doubles):
+                qml.templates.AllSinglesDoubles(weights, wires, hf_state, singles, doubles)
+                return qml.expval(qml.PauliZ(0))
+
+            # Evaluate the QNode for a given set of parameters
+            params = np.random.normal(0, np.pi, len(singles) + len(doubles))
+            circuit(params, hf_state, singles=singles, doubles=doubles)
+    """
+
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, weights, wires, hf_state, singles=None, doubles=None, do_queue=True):
+
+        if len(wires) < 2:
+            raise ValueError(
+                "The number of qubits (wires) can not be less than 2; got len(wires) = {}".format(
+                    len(wires)
+                )
+            )
+
+        if not singles and not doubles:
+            raise ValueError(
+                "'singles' and 'doubles' lists can not be both empty; got singles = {}, doubles = {}".format(
+                    singles, doubles
+                )
+            )
+
+        for d_wires in doubles:
+            if len(d_wires) != 4:
+                raise ValueError(
+                    "Expected entries of 'doubles' to be of size 4; got {} of length {}".format(
+                        d_wires, len(d_wires)
+                    )
+                )
+
+        for s_wires in singles:
+            if len(s_wires) != 2:
+                raise ValueError(
+                    "Expected entries of 'singles' to be of size 2; got {} of length {}".format(
+                        s_wires, len(s_wires)
+                    )
+                )
+
+        shape = qml.math.shape(weights)
```

Comment:
This is optional, but we could consider adding a `shape()` method that the user can employ to obtain the shape of the input parameters, as is done here
https://pennylane.readthedocs.io/en/latest/_modules/pennylane/templates/layers/cv_neural_net.html#CVNeuralNetLayers.shape

Reply:
Actually, I am strongly in favour of `shape()` being added here. It is frustrating from a user point of view for some templates to have this method, and some to not :)

Reply:
If I understand correctly, do we want to add a ``shape()`` method that receives the lists ``singles`` and ``doubles`` and returns ``shape = (len(singles) + len(doubles)) ?
@mariaschuld your instructions here would be valuable :-).

Reply:
Yes, the shape method would have the `@staticmethod` decorator and does not take `self`, which means that all information it needs have to be derived from the argument that `shape()` gets (since you cannot do `self.xyz`).

If the shape can be derived from anything other than `singles` and `doubles`, like the number of qubits or something the user should know about the circuit, I would prefer this - especially if you make the `singles` and `doubles` optional as discussed above!